Don't fire a mousemove event when a modifier key is pressed

aprotyas opened the pull request:
#### f84b2332af6e577f2e2e3abcf8bc9768410c1f52
<pre>
Don&apos;t fire a mousemove event when a modifier key is pressed
<a href="https://bugs.webkit.org/show_bug.cgi?id=16271">https://bugs.webkit.org/show_bug.cgi?id=16271</a>
rdar://81287778

Reviewed by Wenson Hsieh.

Previously, we dispatched mousemove events whenever a modifier key was
pressed. This was used to call the UIClient delegate callback
mouseDidMoveOverElement, which downstream clients could use to modify
various behaviors based on whether or not modifier keys were pressed.

Unfortunately, the act of dispatching a mousemove event meant that
pressing modifier keys could cause unexpected mouse behavior. One notable
example is Slack. When using Slack through Minibrowser, it is possible
to experience spontaneous mouse selection when a cursor was originally
placed behind a completion list, since our hover state gets constantly
re-evaluated.

In this patch, we introduce a mechanism to avoid the need to fire a
mousemove event when modifier keys are pressed. This is achieved by
adding a new message from the UI process to the web process where we
request that a given mouse event&apos;s hit test be performed. With the hit
test result and user data associated with said mouse event, we can
simply call into the UIClient delegate callback mouseDidMoveOverElement,
avoiding the regular mouse event dispatch dance altogether.

* LayoutTests/fast/events/no-mousemove-on-modifier-key-press-expected.txt: Added.
* LayoutTests/fast/events/no-mousemove-on-modifier-key-press.html: Added.

Layout test that verifies pressing the modifier keys does not fire a
MouseMove event.

* Source/WebCore/page/Chrome.h:

Export (and make public) Chrome::getToolTip to access from the web
process.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::getHitTypeForMouseMoveEvent):
(WebCore::EventHandler::getHitTestResultForMouseEvent):
(WebCore::EventHandler::handleMouseMoveEvent):

Introduce two functions getHitTypeForMouseMoveEvent and
getHitTestResultForMouseEvent. The latter strips out some common code
from handleMouseMoveEvent so that it can be called independently to
identify a hitType, whereas the former is a handy way to produce a
hitTestResult without bringing the MouseEventWithHitTestResults type
into web process land.

* Source/WebCore/page/EventHandler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::dispatchMouseDidMoveOverElementAsynchronously):

Introduce a method that communicates the need for user data and a hit
test on a mouse event without actually dispatching said mouse event to
the web process. Once we round trip from the web process, it feeds the
returned data to the client delegate callback mouseDidMoveOverElement.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::viewDidMoveToWindow):
(WebKit::WebViewImpl::fireMouseDidMoveOverElement):
(WebKit::WebViewImpl::postFakeMouseMovedEventForFlagsChangedEvent): Deleted.

Rename postFakeMouseMovedEventForFlagsChangedEvent to
fireMouseDidMoveOverElement to better represent our signaling intent.

fireMouseDidMoveOverElement then communicates with the web page to
receive user data and hit test results before passing that over to the
client delegate callback.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::performHitTestAndGetUserData):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Introduce a new message that receives a mouse event from the UI process
and calls into the page&apos;s event handler and chrome clients to perform a
hit test on the received mouse event. This allows us to circumvent the
usual dance with a dispatched mouse event while returning the user data
and hit test result associated with a mouse event.

Canonical link: <a href="https://commits.webkit.org/264455@main">https://commits.webkit.org/264455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fd5225ad49f0d449066611540d674c7621100e3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7652 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7921 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8105 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9295 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7832 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7664 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9920 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7846 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10704 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7785 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8927 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7048 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9437 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6235 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6971 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/7398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7094 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10540 "5 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7588 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6191 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6922 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6882 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1827 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11133 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7319 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1978,6 +1978,43 @@ bool EventHandler::passMouseMovedEventToScrollbars(const PlatformMouseEvent& eve
 
+OptionSet<HitTestRequest::Type> EventHandler::getHitTypeForMouseMoveEvent(const PlatformMouseEvent& platformMouseEvent, bool onlyUpdateScrollbars)
+{
+    OptionSet<HitTestRequest::Type> hitType { HitTestRequest::Type::Move, HitTestRequest::Type::DisallowUserAgentShadowContent, HitTestRequest::Type::AllowFrameScrollbars };
```
Nit - I think you can fall back on template type deduction here:

```suggestion
    OptionSet hitType { HitTestRequest::Type::Move, HitTestRequest::Type::DisallowUserAgentShadowContent, HitTestRequest::Type::AllowFrameScrollbars };
```

aprotyas replied:
Fixed in ~~https://github.com/WebKit/WebKit/commit/b7607875e50a6089c8262d0f679251a1a5216add~~ https://github.com/WebKit/WebKit/commit/ef38976d78eeb4c553b562487a8a3a6d79ca1f98

whsieh commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1978,6 +1978,43 @@ bool EventHandler::passMouseMovedEventToScrollbars(const PlatformMouseEvent& eve
+HitTestResult EventHandler::getHitTestResultForMouseEvent(const PlatformMouseEvent& platformMouseEvent)
+{
+    auto hitType = getHitTypeForMouseMoveEvent(platformMouseEvent);
+    HitTestRequest request(hitType);
```
Nit - not too sure the local variable helps clarify much here.

```suggestion
    HitTestRequest request(getHitTypeForMouseMoveEvent(platformMouseEvent));
```

aprotyas replied:
Fixed in ~~https://github.com/WebKit/WebKit/commit/b7607875e50a6089c8262d0f679251a1a5216add~~ https://github.com/WebKit/WebKit/commit/ef38976d78eeb4c553b562487a8a3a6d79ca1f98

whsieh commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2022,30 +2059,7 @@ bool EventHandler::handleMouseMoveEvent(const PlatformMouseEvent& platformMouseE
-#endif
-    
+    auto hitType = getHitTypeForMouseMoveEvent(platformMouseEvent, onlyUpdateScrollbars);
     HitTestRequest request(hitType);
```
Same here — maybe just this?

```suggestion
    HitTestRequest request(getHitTypeForMouseMoveEvent(platformMouseEvent, onlyUpdateScrollbars));
```

aprotyas replied:
Fixed in ~~https://github.com/WebKit/WebKit/commit/b7607875e50a6089c8262d0f679251a1a5216add~~ https://github.com/WebKit/WebKit/commit/ef38976d78eeb4c553b562487a8a3a6d79ca1f98

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3180,6 +3180,13 @@ void WebPageProxy::handleMouseEvent(const NativeWebMouseEvent& event)
+void WebPageProxy::shouldFireMouseDidMoveOverElement(const NativeWebMouseEvent& event)
+{
+    sendWithAsyncReply(Messages::WebPage::PerformHitTestAndGetUserData { event }, [this, protectedThis = Ref { *this }](WebHitTestResultData&& hitTestResult, OptionSet<WebEventModifier>&& modifiers, UserData&& userData) {
+        this->mouseDidMoveOverElement(WTFMove(hitTestResult), WTFMove(modifiers).toRaw(), WTFMove(userData));
```
Hm…I don't think it's technically correct to call `.toRaw()` after moving the modifiers, though I think in this case it shouldn't make a difference since it's basically just a cast. Might be cleaner to simply pass `OptionSet<WebEventModifier>` by value, since it's just a byte anyways.

```suggestion
        this->mouseDidMoveOverElement(WTFMove(hitTestResult), modifiers.toRaw(), WTFMove(userData));
```

aprotyas replied:
Yes you're right, forgot to remove the move when I realized `WebPageProxy::mouseDidMoveOverElement` expects an opaque modifier flag.

aprotyas replied:
I'll just pass the `OptionSet` by value instead

pxlcoder replied:
Not sure if this is what you meant, but can we also change the `uint32_t` to an `OptionSet` in the `messages.in` file, if it doesn't increase the patch size too much? Otherwise, we should probably do it in a follow-up.

`MouseDidMoveOverElement(struct WebKit::WebHitTestResultData hitTestResultData, uint32_t modifiers, WebKit::UserData userData)` in `WebPageProxy.messages.in`.

aprotyas replied:
> Not sure if this is what you meant, but can we also change the uint32_t to an OptionSet in the messages.in file, if it doesn't increase the patch size too much? Otherwise, we should probably do it in a follow-up.

Yes this is what I meant. Let me look into this idea. I know at some point in the roundtrip from the web process the callstack starts speaking in terms of `uint32_t` modifierFlags instead of `OptionSet<>` but I'm not sure why... hopefully is a simple fix.

aprotyas replied:
Ah, so `EventHandler` calls into `WebCore::Chrome::mouseDidMoveOverElement`, which is why we can't `OptionSet<WebEventModifier>` everything.

We could adopt `OptionSet<PlatformEvent::Modifier>` in liue, but that's out of scope of this PR so I'll just add a TODO to that effect.

aprotyas replied:
Changed to pass the `OptionSet` by value in ~~https://github.com/WebKit/WebKit/commit/b7607875e50a6089c8262d0f679251a1a5216add~~ https://github.com/WebKit/WebKit/commit/ef38976d78eeb4c553b562487a8a3a6d79ca1f98

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3180,6 +3180,13 @@ void WebPageProxy::handleMouseEvent(const NativeWebMouseEvent& event)
         processNextQueuedMouseEvent();
 }
 
+void WebPageProxy::shouldFireMouseDidMoveOverElement(const NativeWebMouseEvent& event)
```
Hm… `shouldFireMouseDidMoveOverElement` is a bit of an odd name for this function, since it reads like a method that should return a `bool` with no other side effects.

Perhaps something like `fireMouseDidMoveOverElementAfterHitTest()` or `dispatchMouseDidMoveOverElementAsynchronously()`?

aprotyas replied:
Okay yes I see how the name is misleading. I'll just rename to `dispatchMouseDidMoveOverElementAsynchronously` with the idea that the callee needn't know why a hit test is needed for this dispatch.

aprotyas replied:
Fixed in ~~https://github.com/WebKit/WebKit/commit/b7607875e50a6089c8262d0f679251a1a5216add~~ https://github.com/WebKit/WebKit/commit/ef38976d78eeb4c553b562487a8a3a6d79ca1f98

hortont424 commented on Source/WebKit/UIProcess/mac/WebViewImpl.h:
```diff
@@ -751,6 +749,8 @@ ALLOW_DEPRECATED_DECLARATIONS_END
     void nativeMouseEventHandler(NSEvent *);
     void nativeMouseEventHandlerInternal(NSEvent *);
     
+    void fireMouseDidMoveOverElement(NSEvent *);
```
Not exactly the right name, since it doesn't call mouseDidMoveOverElement. Maybe `request` or `schedule` or something that indicates it's a later thing?

Alternatively, less of a change from the old name? postFakeMouseDidMoveOverElementCallbackForFlagsChangedEvent? Quite wordy, maybe not that :P

aprotyas replied:
To avoid a super wordy name.. I'll go with `scheduleMouseDidMoveOverElement`!

aprotyas replied:
https://github.com/WebKit/WebKit/pull/14221/commits/c88025563e2436195514cb0c0a22883be08bf627

hortont424 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3395,6 +3396,26 @@ void WebPage::mouseEvent(const WebMouseEvent& mouseEvent, std::optional<Vector<S
     revokeSandboxExtensions(mouseEventSandboxExtensions);
 }
 
+void WebPage::performHitTestAndGetUserData(const WebMouseEvent& event, CompletionHandler<void(WebHitTestResultData&&, OptionSet<WebEventModifier>&&, UserData&&)>&& completionHandler)
```
maybe `performHitTestForMouseEvent`?

aprotyas replied:
Renamed to `performHitTestForMouseEvent` in https://github.com/WebKit/WebKit/pull/14221/commits/c88025563e2436195514cb0c0a22883be08bf627

whsieh commented on LayoutTests/fast/events/no-mousemove-on-modifier-key-press.html:
```diff
@@ -0,0 +1,43 @@
+            UIHelper.keyDown("k", [modifierKey]);
+            debug('Dispatched KeyDown event with modifier ' + modifierKey);
+        });
+    }
```
`UIHelper.keyDown` is asynchronous on iOS, where we use `UIScriptController`. We should probably await these calls:

```suggestion
    async function fireKeyDownEventsWithModifiers() {
        ['metaKey', 'altKey', 'shiftKey', 'ctrlKey'].forEach(modifierKey => {
            await UIHelper.keyDown("k", [modifierKey]);
            debug('Dispatched KeyDown event with modifier ' + modifierKey);
        });
    }
```

aprotyas replied:
Asyncified the test in https://github.com/WebKit/WebKit/pull/14221/commits/c88025563e2436195514cb0c0a22883be08bf627

whsieh commented on LayoutTests/fast/events/no-mousemove-on-modifier-key-press.html:
```diff
@@ -0,0 +1,43 @@
+    window.addEventListener('mousemove', () => testFailed("MouseMove event detected"));
+
+    if (window.testRunner)
+        fireKeyDownEventsWithModifiers();
```
We'll want to `await` here as well once we make the above change.

aprotyas replied:
Yup, asyncified the test in https://github.com/WebKit/WebKit/pull/14221/commits/c88025563e2436195514cb0c0a22883be08bf627

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3395,6 +3396,26 @@ void WebPage::mouseEvent(const WebMouseEvent& mouseEvent, std::optional<Vector<S
 
+void WebPage::performHitTestAndGetUserData(const WebMouseEvent& event, CompletionHandler<void(WebHitTestResultData&&, OptionSet<WebEventModifier>&&, UserData&&)>&& completionHandler)
+{
+    auto* localMainFrame = dynamicDowncast<WebCore::LocalFrame>(corePage()->mainFrame());
```
Nit - `RefPtr` here, to guarantee that the pointer remains valid over its entire scope.

aprotyas replied:
Good idea -- fixed in https://github.com/WebKit/WebKit/pull/14221/commits/c88025563e2436195514cb0c0a22883be08bf627

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3395,6 +3396,26 @@ void WebPage::mouseEvent(const WebMouseEvent& mouseEvent, std::optional<Vector<S
+    if (!localMainFrame || !localMainFrame->view())
+        return;
+
+    auto hitTestResult = Ref(*localMainFrame)->eventHandler().getHitTestResultForMouseEvent(platform(event));
```
(And then you won't need the `Ref` here with the `RefPtr` above)

whsieh reviewed (approved):
LGTM with Tim's comments too.

aprotyas committed:
Don't fire a mousemove event when a modifier key is pressed
https://bugs.webkit.org/show_bug.cgi?id=16271
rdar://81287778

Reviewed by Wenson Hsieh.

Previously, we dispatched mousemove events whenever a modifier key was
pressed. This was used to call the UIClient delegate callback
mouseDidMoveOverElement, which downstream clients could use to modify
various behaviors based on whether or not modifier keys were pressed.

Unfortunately, the act of dispatching a mousemove event meant that
pressing modifier keys could cause unexpected mouse behavior. One notable
example is Slack. When using Slack through Minibrowser, it is possible
to experience spontaneous mouse selection when a cursor was originally
placed behind a completion list, since our hover state gets constantly
re-evaluated.

In this patch, we introduce a mechanism to avoid the need to fire a
mousemove event when modifier keys are pressed. This is achieved by
adding a new message from the UI process to the web process where we
request that a given mouse event's hit test be performed. With the hit
test result and user data associated with said mouse event, we can
simply call into the UIClient delegate callback mouseDidMoveOverElement,
avoiding the regular mouse event dispatch dance altogether.

* LayoutTests/fast/events/no-mousemove-on-modifier-key-press-expected.txt: Added.
* LayoutTests/fast/events/no-mousemove-on-modifier-key-press.html: Added.

Layout test that verifies pressing the modifier keys does not fire a
MouseMove event.

* Source/WebCore/page/Chrome.h:

Export (and make public) Chrome::getToolTip to access from the web
process.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::getHitTypeForMouseMoveEvent):
(WebCore::EventHandler::getHitTestResultForMouseEvent):
(WebCore::EventHandler::handleMouseMoveEvent):

Introduce two functions getHitTypeForMouseMoveEvent and
getHitTestResultForMouseEvent. The latter strips out some common code
from handleMouseMoveEvent so that it can be called independently to
identify a hitType, whereas the former is a handy way to produce a
hitTestResult without bringing the MouseEventWithHitTestResults type
into web process land.

* Source/WebCore/page/EventHandler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::dispatchMouseDidMoveOverElementAsynchronously):

Introduce a method that communicates the need for user data and a hit
test on a mouse event without actually dispatching said mouse event to
the web process. Once we round trip from the web process, it feeds the
returned data to the client delegate callback mouseDidMoveOverElement.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::viewDidMoveToWindow):
(WebKit::WebViewImpl::fireMouseDidMoveOverElement):
(WebKit::WebViewImpl::postFakeMouseMovedEventForFlagsChangedEvent): Deleted.

Rename postFakeMouseMovedEventForFlagsChangedEvent to
fireMouseDidMoveOverElement to better represent our signaling intent.

fireMouseDidMoveOverElement then communicates with the web page to
receive user data and hit test results before passing that over to the
client delegate callback.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::performHitTestAndGetUserData):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Introduce a new message that receives a mouse event from the UI process
and calls into the page's event handler and chrome clients to perform a
hit test on the received mouse event. This allows us to circumvent the
usual dance with a dispatched mouse event while returning the user data
and hit test result associated with a mouse event.

Canonical link: https://commits.webkit.org/264455@main
